File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
const fs = require('fs');
const input = fs.readFileSync(0, 'utf-8').toString().trim().split('\n');
const [N, M] = input[0].split(' ').map(Number);
const maze = input.slice(1).map(row => row.split('').map(x => +x === 1 ? 0 : -1));

maze[0][0] = 1;

const queue = [[0, 0]];
let index = 0;
const dn = [1, 0, -1, 0];
const dm = [0, 1, 0, -1];

while (index < queue.length) {
    const [cn, cm] = queue[index++];
    
    for (let i = 0; i < 4; i++) {
        const nn = cn + dn[i];
        const nm = cm + dm[i];
        
        if (nn < 0 || nn >= N) continue;
        if (nm < 0 || nm >= M) continue;
        if (maze[nn][nm] !== 0) continue;
        
        maze[nn][nm] = maze[cn][cm] + 1;
        
        if (nn === N - 1 && nm === M - 1) {
            console.log(maze[nn][nm]);
            process.exit();
        }
        
        queue.push([nn, nm]);
    }
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
from collections import deque

N, M = map(int, input().split())
visited = [[0 for _ in range(M)] for _ in range(N)]
maze = [input() for _ in range(N)]
queue = deque([[0, 0]])
di = [1, 0, -1, 0]
dj = [0, 1, 0, -1]
visited[0][0] = 1
repeat_flag = True

while repeat_flag and len(queue) > 0:
    ci, cj = queue.popleft()
    
    for k in range(4):
        ni = ci + di[k]
        nj = cj + dj[k]

        if (ni >= 0 and nj >= 0 and ni < N and nj < M
           and visited[ni][nj] == 0 and maze[ni][nj] == '1'):
            visited[ni][nj] = visited[ci][cj] + 1
            queue.append([ni, nj])
            
            if ni == N-1 and nj == M-1:
                repeat_flag = False

print(visited[N-1][M-1])